GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1257,15 +1319,104 @@ private string GetNewProjectedShaForPlaceholder(string path)
+                else
+                {
+                    childRelativePath = relativeFolderPath + Path.DirectorySeparatorChar + childEntry.Name.GetString();
+                }
+
+                // TODO(Mac): Issue #245, handle failures of WritePlaceholderDirectory and WritePlaceholderFile         
+                if (childEntry.IsFolder)
+                {
+                    if (!existingFolderPlaceholders.Contains(childRelativePath))
+                    {
+                        this.fileSystemVirtualizer.WritePlaceholderDirectory(childRelativePath);                    
+                        updatedPlaceholderList.TryAdd(
+                            childRelativePath, 
+                            new PlaceholderListDatabase.PlaceholderData(childRelativePath, PlaceholderListDatabase.PartialFolderValue));
+                    }
+                }
+                else
+                {
+                    if (!updatedPlaceholderList.ContainsKey(childRelativePath))
+                    {
+                        FileData childFileData = childEntry as FileData;
+                        string sha = childFileData.Sha.ToString();
+
+                        this.fileSystemVirtualizer.WritePlaceholderFile(childRelativePath, childFileData.Size, sha);
+                        updatedPlaceholderList.TryAdd(
+                            childRelativePath, 
+                            new PlaceholderListDatabase.PlaceholderData(childRelativePath, sha));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes the folder placeholder from disk if it's empty.
+        /// </summary>
+        /// <returns>
+        /// <c>true</c>If the folder placeholder was deleted 
+        /// <c>false</c>If RemoveFolderPlaceholderIfEmpty did not attempt to remove the folder placeholder
+        /// </returns>
+        /// <remarks>
+        /// If the platform expands on enumeration the folder will only be removed if it's not in the projection
+        /// </remarks>
+        private bool RemoveFolderPlaceholderIfEmpty(
             PlaceholderListDatabase.PlaceholderData placeholder,
-            ConcurrentBag<PlaceholderListDatabase.PlaceholderData> updatedPlaceholderList,
+            Action<PlaceholderListDatabase.PlaceholderData> addPlaceholderToUpdatedPlaceholders,
             ConcurrentHashSet<string> folderPlaceholdersToKeep)
         {
             if (folderPlaceholdersToKeep.Contains(placeholder.Path))
             {
-                updatedPlaceholderList.Add(placeholder);
-                return;
+                addPlaceholderToUpdatedPlaceholders(placeholder);
+                return false;
+            }
+
+            if (GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories)
+            {
+                // If enumeration expands directories we should leave folder placeholders
+                // that are still in the projection on disk (even if they are empty).

```

Comment:
Why do we leave them even if they are empty? How then do empty directories after a checkout get cleaned up?

Reply:
> Why do we leave them even if they are empty?

I think should update this comment.  "Empty" here means that on the physical disk the folder is empty (e.g. because the folder has not been expanded yet).

> How then do empty directories after a checkout get cleaned up?

If they are empty from git's perspective they will *not* be in the projection, and so `TryGetOrAddFolderDataFromCache` will return false and VFSForGit will proceed with the delete of the folder.

(On Windows we can't make this "in projection" check because we also have to worry about tombstones that might be on disk)